internal/ls/userpreferences.go

```diff
@@ -209,3 +347,287 @@ func (p *UserPreferences) ModuleSpecifierPreferences() modulespecifiers.UserPref
 		AutoImportSpecifierExcludeRegexes: p.AutoImportSpecifierExcludeRegexes,
 	}
 }
+
+// ------ Parsing Config Response -------
+
+// returns non-nil if should break loop
+func (p *UserPreferences) Parse(item any) *UserPreferences {
+	if item == nil {
+		// continue
+	} else if config, ok := item.(map[string]any); ok {
+		p.parseWorker(config)
+	} else if item, ok := item.(*UserPreferences); ok {
+		// case for fourslash
+		return item.CopyOrDefault()
+	}
+	return nil
+}
+
+func (p *UserPreferences) parseWorker(config map[string]interface{}) {
+	// Process unstable preferences first so that they do not overwrite stable properties
+	if unstable, ok := config["unstable"]; ok {
+		// unstable properties must be named the same as userPreferences
+		p.parseAll(unstable)
+	}
+	for name, values := range config {
+		switch name {
+		case "unstable":
+			continue
+		case "inlayHints":
+			p.parseInlayHints(values)
+		case "suggest":
+			p.parseSuggest(values)
+		case "preferences":
+			p.parsePreferences(values)
+		case "tsserver":
+			// !!!
+		case "tsc":
+			// !!!
+		case "experimental":
+			// !!!
+		default:
+			p.set(name, values)
+		}
+	}
+}
+
+func (p *UserPreferences) parseAll(prefs any) {
+	prefsMap, ok := prefs.(map[string]any)
+	if !ok {
+		return
+	}
+	for name, value := range prefsMap {
+		p.set(name, value)
+	}
+}
+
+func (p *UserPreferences) parseInlayHints(prefs any) {
+	inlayHintsPreferences, ok := prefs.(map[string]any)
+	if !ok {
+		return
+	}
+	for name, value := range inlayHintsPreferences {
+		if v, ok := value.(map[string]any); ok {
+			// vscode's inlay hints settings are nested objects with "enabled" and other properties
+			switch name {
+			case "parameterNames":
```

Comment:
We are case sensitive in some places, but not in others?

Reply:
case sensitive is for vscode specifically